pkg/controller/remotemachineset/gcpactuator.go

```diff
@@ -142,3 +195,145 @@ func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldL
 		return "", errors.New(msg)
 	}
 }
+
+// obtainLease uses the Hive MachinePoolNameLease resource to obtain a unique, single character
+// for use in the name of the machine pool. We are severely restricted on name lengths on GCP
+// and effectively have one character of flexibility with the naming convention originating in
+// the installer.
+func (a *GCPActuator) obtainLease(pool *hivev1.MachinePool, cd *hivev1.ClusterDeployment, logger log.FieldLogger) (string, bool, error) {
+	logger.Debugf("obtaining lease for pool: %s", pool.Name)
+
+	leases := &hivev1.MachinePoolNameLeaseList{}
+	err := a.client.List(context.TODO(), leases, client.InNamespace(pool.Namespace),
+		client.MatchingLabels(map[string]string{constants.ClusterDeploymentNameLabel: cd.Name}))
+	if err != nil {
+		return "", false, err
+	}
+	logger.Debugf("found %d leases in namespace", len(leases.Items))
+	for _, l := range leases.Items {
+		logger.Debugf("checking lease: %s", l.Name)
+		for _, ownerRef := range l.OwnerReferences {
+			logger.Debugf("checking owner ref: %v", ownerRef)
+			if ownerRef.APIVersion == "hive.openshift.io/v1" &&
+				ownerRef.Kind == "MachinePool" &&
+				ownerRef.Name == pool.Name {
+
+				logger.Debugf("machine pool already has lease: %s", l.Name)
+
+				// Ensure the lease name is in the format we expect, we know everything up to
+				// the last character.
+				expectedLeasePrefix := fmt.Sprintf("%s-", cd.Spec.ClusterMetadata.InfraID)
+				if l.Name[0:len(l.Name)-1] != expectedLeasePrefix {
+					return "", false, fmt.Errorf("lease %s did not match expected lease name format (%s[CHAR])", l.Name, expectedLeasePrefix)
```

Comment:
Should Hive do something to resolve this?

Reply:
I'm not sure what it would be. I don't think I'd dare delete it and let it create a new one as we might nuke a machine set. It felt like a pretty fringe/unexpected thing to happen, and a fair candidate to refuse to reconcile and require manual intervention from an admin who can figure out what happened.